### PR TITLE
Improve accuracy of CalcTextWidth if selected font has kerning pairs

### PR DESCRIPTION
**Pull request by kbuffington**
I discovered that if a font has kerning pairs and the text you wish to get the width of contains one (or more) of those pairs the value of `CalcTextWidth` will be off for each pair in the text. I couple **Yo**u's or **To**'s and things get really bad really quickly (the cursor and highlighted text should be directly adjacent to the previous Y):

bad calc

See discussion [here](https://stackoverflow.com/questions/67683932/how-to-force-gettextextentpoint32-to-take-into-account-kerning?noredirect=1#comment119648000_67683932) where the suggestion was made to switch to using `DrawText` with `DT_CALCRECT`.

I made the change and now things look good again:

image

In the process of investigating I also discovered that despite the documentation, `GdiDrawText` does _not_ return:
```
     * @return {Array<number>}
     *     index | meaning <br>
     *     [0] left   (DT_CALCRECT) <br>
     *     [1] top    (DT_CALCRECT) <br>
     *     [2] right  (DT_CALCRECT) <br>
     *     [3] bottom (DT_CALCRECT) <br>
     *     [4] characters drawn
```
even if DT_CALCRECT flag is set. I had no idea how you wanted to handle that, so I left it as is.

### CODE REVIEW

**Comment by kbuffington** [comment]
@marc2k3 `CalcTextWidth` will be affected in JSP as well, although over there at least `GdiDrawText` does not claim to return any values :)

**Comment by marc2k3** [comment]
This new code doesn't like large strings/being called repeatedly as is done by the EstimateLineWrap function. It freezes fb2k indefinitely until you kill it with task manager. Try the last.fm bio sample.

**Comment by kbuffington** [comment]
Yeah, I'm seeing the same thing. For some reason I've never been able to attach to FSM's dll's to debug exactly what's going wrong. I can understand it being slow, but doesn't make sense why it would lock up FB.

Simple solution would be to have EstimateLineWraps's recursive functions continue to always call GetTextExtentPoint32 I suppose.

**Comment by kbuffington** [comment]
Should be safe now as it only calls DrawText from CalcTextWidth, and in all other cases uses the less accurate version which should be perfectly fine.

**Comment by marc2k3** [comment]
This is buggy if the text contains ampersands. You need to combine `DT_PREFIX` with `DT_CALCRECT` when using `DrawText`.

**Comment by kbuffington** [comment]
D'oh. Should have caught that. Also added a check to skip this block if text.length is 1 or less. Kerning pairs don't matter if you can't have a kerning pair.

**Comment by TheQwertiest** [comment]
@kbuffington thanks for your contribution!
There are a few things that need to be changed before it can be merged though:
- This is a breaking change, hence it must be moved under a separate argument (e. g. use_exact or smth) and should not be enabled by default.
- This new argument must be added to docs.
- DT_SINGLELINE should be added to DrawText call.

File: foo_spider_monkey_panel/utils/text_helpers.cpp [hunk 1]
```diff
@@ -111,11 +111,22 @@ size_t get_text_height( HDC hdc, std::wstring_view text )
     return static_cast<size_t>( size.cy );
 }
 
-size_t get_text_width( HDC hdc, std::wstring_view text )
+size_t get_text_width( HDC hdc, std::wstring_view text, bool accurate )
 {
     SIZE size;
```
Comment (TheQwertiest) [inline foo_spider_monkey_panel/utils/text_helpers.cpp:116]: Size is only used in the second branch of if, so it should be moved there

File: foo_spider_monkey_panel/utils/text_helpers.cpp [hunk 2]
```diff
@@ -111,11 +111,22 @@ size_t get_text_height( HDC hdc, std::wstring_view text )
...
+        RECT rc_calc{ 0, 0, 0, 0 };
+        DrawText( hdc, text.data(), -1, &rc_calc, DT_CALCRECT | DT_NOPREFIX );
+        return rc_calc.right;
+    }
+    else
+    {
+        // TODO: add error checks
+        GetTextExtentPoint32( hdc, text.data(), static_cast<int>( text.size() ), &size );
+    }
     return static_cast<size_t>( size.cx );
```
Comment (TheQwertiest) [inline foo_spider_monkey_panel/utils/text_helpers.cpp:130]: move return into if branch

**Comment by TheQwertiest** [comment]
@kbuffington can you fix the PR, plz? So that I could include it in the release (which I'm planning to ship soon).

**Comment by kbuffington** [comment]
Yeah, I'll try to make some time for this. I'm not completely sure what you meant by: "DT_SINGLELINE should be added to DrawText call." but I haven't looked at the code in a month so maybe it'll make more sense when I do.

**Comment by TheQwertiest** [comment]
DT_SINGLELINE single line causes method to ignore new lines (i.e. `\n`), this way it would work the same as GetTextExtentPoint32

**Comment by kbuffington** [comment]
After looking at the code I understand what you were referencing and will add that.

**Comment by kbuffington** [comment]
Believe I've addressed everything, retested with new optional parameters and it all seems good.

**Comment by TheQwertiest** [comment]
Awesome, thanks!

**Pull request merged**